### ISSUE TITLE
onMessage/postMessage WebView Communication not work on Expo 32 but work on Expo 31 or below

### ISSUE THREAD

**Issue by imben1109** [expo/expo#3168]
I am using Expo to integrate webview. I use onMessage/postMessage WebView Communication for WebView Communication.

onMessage/postMessage WebView Communication not work on Expo 32 but work on Expo 31 or below. 

```
import React, { Component } from 'react';
import { View, StyleSheet, WebView, Text , Alert} from 'react-native';
import { Constants } from 'expo';

export default class App extends Component {
  render() {
    return (
      <View style={{flex: 1, paddingTop: Constants.statusBarHeight}}>
        <WebViewAutoHeight />
      </View>
    );
  }
}

class WebViewAutoHeight extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      receivedMessageCount:0,
      receivedMessages: [],
    }
  }
  
  addMessage(message){
    Alert.alert("Something", "Something");
    this.setState({
      receivedMessages: [...this.state.receivedMessages, message]
    })
  }

  render() {
    return (
      <View style={{flex:1, width:'100%'}}>
        <Text style={{backgroundColor:'black', color:'white', fontSize:20}}>{this.state.receivedMessages.length} Messages from webview</Text>
        <Text style={{backgroundColor:'black', color:'white', fontSize:14}}>
          {this.state.receivedMessages.join('\n')}
        </Text>
        <WebView
          source={{html:`
            <html>
              <head>
                <script>
                  function test(){
                    window.postMessage("abc", '*')
                    alert("test");
                  }
                </script>
              </head>
              <body>
                <button onClick="test()">Test7</button>
              </body>
            </html>
          `}}
          scrollEnabled={false}
          onMessage={(evt)=>{
            this.addMessage(evt.nativeEvent.data);
          }}
          javaScriptEnabled={true}
        />

      </View>
    );
  }
}
```

**Comment by schellack** [comment]
I have the same problem, though only on iOS. Here is a simple snack you can also use as another example: [https://snack.expo.io/@salus/posting-from-webview-to-rn](https://snack.expo.io/@salus/posting-from-webview-to-rn)

Note that in my example, I am calling an external url as the WebView source. This is to show that the web server is returning a 404, because the `window.postMessage` call is not posting back to the web view (or the web view is not accepting it properly). Instead, the web view is attempting to navigate to `ReactABI32_0_0-js-navigation://postMessage`. That's less apparent when running inline html in the web view.

If you downgrade the snack to sdk version 31 and reload the app, the problem goes away.

**Comment by tsapeta** [comment]
Just submitted a new Expo Client build for Apple review 😎 An update will be available once they approve it.

**Comment by lgandecki** [comment]
@schellack what steps do you take to "downgrade and reload"?

Is this released somewhere, or is there a way to use this fix? 
I wasted a few very frustrating hours today trying to understand what am I doing wrong... :( I tried the 32.0.1 but it doesn't seem to be fixed. 
I also tried the 31.0.6 and 31.0.0 and they also produce the same result. I changed the expo version and the react-native sdk. to sdk-31. Not sure if this is beacuse of some cache. I removed and reinstalled node_modules, and run expo with expo start -c. I Erased all content and settings on the simulator, and this is still happening.

Don't want to complain - I understand that this is a free and open source project and I appreciate that. I maintain a few of those by myself. Nonetheless, I would feel bad leaving a regression that brakes applications for weeks on end, especially if the fix is known and merged..

**Comment by tsapeta** [comment]
@lgandecki besides downgrading `expo` package to `31.0.6`, you should also set `sdkVersion` in your `app.json` config to `31.0.0`, otherwise Expo Client will continue to use the native code for SDK32 which has this bug.

We also apologize that the fix is not available to use yet, but it happened for reasons beyond our control 😞 We're really working hard to get it out asap.

**Comment by lgandecki** [comment]
Thank you for an immediate response, and on Sunday! 💯 
Yes, app.json did the trick and the "ReactABI32_0_0-js-navigation://postMessage" issue disappeared.

The functionality still doesn't work but I probably messed something up while trying to figure out what's wrong, so it's probably on my side. 

thanks again!

**Comment by johnnymajor** [comment]
After downgrading to 31 I am now getting an error : _Unable to resolve "expo-app-auth" from "node_modules/expo/build/Expo.js_

I removed the node_modules and updated the app.json to "31.0.0" & package.json to "31.0.6". Any ideas on what I have missed on the downgrade?

Thanks

**Comment by lgandecki** [comment]
Any ideas when this will be published? I just realized that the expo 31 uses the old version of webview which causes iOS really unusable (half a second lag after a click, among other things). I've actually rearchitectured my app from ejected back to non-ejected once 32 was released (the only reason for ejection was to use a custom, performant, webview). I would really like to avoid migrating it again to ejected because of this bug, but I do need to release this in a week. :-(

**Comment by lgandecki** [comment]
Can we have any update here? Are you guys still waiting on Apple Review? @tsapeta please advise, I need to make a decision about release this week.

**Comment by tsapeta** [comment]
Sorry @lgandecki for not responding. Yes, we're still struggling with their review, so I'm afraid that if you want this bug fixed in Expo Client you would have to downgrade to SDK31 😔
Anyway, the fix for this has been deployed to turtles, so standalone apps should work as expected - I hope that's enough for you 🙂

**Comment by lgandecki** [comment]
oh, interesting. So if I build IPA / APK they will work correctly? it's just a matter of not being able to test with expo app? I could definitely live with that :)  
thanks for the reply!

**Comment by lgandecki** [comment]
Yup, thanks for the confirmation. I see the same behavior. It makes the development a bit more difficult (can't use expo for testing anymore..) but besides that, it works, which is most important for me at the moment!

**Comment by mrikirill** [comment]
@lgandecki with my workflow, I'm keeping my project spited for 2 parts or projects:
1 Part: 'Expo detached' when I work with native modules 
2 Part: 'Pure Expo' with a method which disables all non-expo native modules without crashes. That allow me quickly work with JS and UI thinks.

**Comment by lgandecki** [comment]
Interesting perspective, thank you for this input, I might try it..

**Comment by rdibona** [comment]
Backing off to 31 worked for me... Hopefully the fix for 32 comes out soon!

**Comment by lysontrung** [comment]
@tsapeta 
Thanks for the help
How do I know when the problem is resolve ?
In the next version (version 33) ???

**Comment by tsapeta** [comment]
@lysontrung,
It is already solved and deployed to Turtles, so standalone apps are no longer affected. Also, it should work in ejected apps if you use the latest tag for iOS.
Unfortunately we're still not able to release any Expo Client updates to the App Store, so possibly we will update it when we release SDK33.

**Comment by mehmetnyarar** [comment]
Glad to see that the [`react-native-webview`](https://github.com/expo/expo/milestone/22) will be added to SDK33. As far as I see, I have 2 options for now:

1. Downgrade to SDK31
2. Eject and use `react-native-webview`

Just wanna be sure: is there any other solution rather than these two?

If not, I really don't want to downgrade, but I'm wondering whether I can un-eject safely and continue to use `react-native-webview` with SDK33?

**Comment by MehmetKaplan** [comment]
> Unfortunately we're still not able to release any Expo Client updates to the App Store, so possibly we will update it (with this fix) when we release SDK33.

@tsapeta, first thanks for the responses. I searched but could not find the roadmap (though in https://expo.canny.io I see comments like "might be 1-2 months").

Do you have any idea on that? For my case depending on this, probably I'll also need to go back to v31.

**Comment by MehmetKaplan** [comment]
> ...the roadmap...

For anyone who may be interested: https://github.com/expo/expo/milestone/22

**Comment by Elijen** [comment]
Since nobody mentioned yet: Another workaround is to use `useWebKit={true}` option. If you choose to do that you might want to check this: https://facebook.github.io/react-native/blog/2018/08/27/wkwebview

**Comment by MehmetKaplan** [comment]
> Another workaround is to use `useWebKit={true}` option.

Just for an info. Unfortunately in my case this did not work.

`
			<View style={styles.container}>
				<WebView
					ref={webview => { this.webview = webview; }}
					onLoad={() => {
						if ( this.state.webViewLoaded ) return;
						this.setState({ webViewLoaded: true });
					}}
					source={{
						//html: "<html><body>DEFAULT TEXT</body></html>"
						html: "DEFAULT TEXT2"
					}}
					injectedJavaScript={this.injectjs()}
					javaScriptEnabled={true}
					style={styles.webview}
					onMessage={this.state.webViewLoaded ? this.handleMessage : null}
					useWebKit={true}
				/>
			</View>
`

**Comment by mylex** [comment]
Yeah, Downgrading to SDK 31 working for me. 
But, sometimes it shows the previous splash screen like eyeblink while navigate to another page and back to the previous page. 
And become Ok automatically.

**Comment by akagaeng** [comment]
> Since nobody mentioned yet: Another workaround is to use `useWebKit={true}` option. If you choose to do that you might want to check this: https://facebook.github.io/react-native/blog/2018/08/27/wkwebview

It works for me on "SDK 32"!

**Comment by ru88s** [comment]
> use useWebKit={true}

It moved on SDK 32! Thank you very much!

**Comment by ru88s** [comment]
I was downgraded to 31 because I had trouble.

**Comment by lysontrung** [comment]
> @lysontrung,
> It is already solved and deployed to Turtles, so standalone apps are no longer affected. Also, it should work in ejected apps if you use the latest tag for iOS.
> Unfortunately we're still not able to release any Expo Client updates to the App Store, so possibly we will update it (with this fix) when we release SDK33.

Hi, i was upgraded SDK33, but webview work not well,
In my case, I got this respond after use onMessage function : %252C%25C4%2590%25C3%25A3%2520login%2520th%25C3%25A0nh%2520c%25C3%25B4ng"
that's not exactly the respond that server wants to return to me

@tsapeta

**Comment by lysontrung** [comment]
> > @lysontrung,
> > It is already solved and deployed to Turtles, so standalone apps are no longer affected. Also, it should work in ejected apps if you use the latest tag for iOS.
> > Unfortunately we're still not able to release any Expo Client updates to the App Store, so possibly we will update it (with this fix) when we release SDK33.
> 
> Hi, i was upgraded SDK33, but webview work not well,
> In my case, I got this respond after use onMessage function : %252C%25C4%2590%25C3%25A3%2520login%2520th%25C3%25A0nh%2520c%25C3%25B4ng"
> that's not exactly the respond that server wants to return to me
> 
> @tsapeta
this is a respond i got if use SDK31 or below:
Đã login thành công
i don't know sequence %{number}, what does it mean? and how to resolve it ???

**Comment by tsapeta** [comment]
@lysontrung,
First of all, I'd like to remind that WebView included in Expo actually come from [react-native-webview](https://github.com/react-native-community/react-native-webview) so if you have any issues it might be worth to take a look at this repo.
This thread is also not a good place to post such issue, it's completely not related to the main problem we talked about - please create a new issue instead.
Just for a quick solution - it looks like your response is encoded and you need to use `decodeURIComponent` function to decode it. Weird thing is that, it was encoded twice, so you'll have to use `decodeURIComponent` twice 🤔
